codec-http/src/main/java/io/netty/handler/codec/http/HttpHeadersBuilder.java

```diff
@@ -0,0 +1,308 @@
+/*
+ * Copyright 2023 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.handler.codec.http;
+
+import io.netty.handler.codec.DefaultHeaders.NameValidator;
+import io.netty.handler.codec.DefaultHeaders.ValueValidator;
+
+import static io.netty.util.internal.ObjectUtil.checkNotNull;
+
+/**
+ * A builder of {@link HttpHeadersFactory} instances, that itself implements {@link HttpHeadersFactory}.
+ * The builder is immutable, and every {@code with-} method produce a new, modified instance.
+ * <p>
+ * The default builder you most likely want to start with is {@link HttpHeaders#DEFAULT_HEADER_FACTORY}.
+ */
+public class HttpHeadersBuilder implements HttpHeadersFactory {
```

Comment:
1. What is the value of keeping it non-final?
2. I personally find it confusing that factory has a builder that doesn't have a method "build", but is a factory by itself 🙈. Did you consider having static methods on `HttpHeadersFactory` instead of creating factory-builder-factory cycle? If the reason is backward compatibility with older JDK versions, consider creating `HttpHeadersFactories` as a final utility class with static methods to create all range of factories.